<a href="https://colab.research.google.com/github/tohnperfect/DPDM2024/blob/main/Frequent_Patterns_(Association_Rules)_(dpdm24).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import os

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
path = '/content/drive/MyDrive/dpdm24_DATA'

In [4]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [5]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [10]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [16]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

## Association Rules

In [11]:
!ls

drive  sample_data


In [12]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5953 sha256=b9809bbcca977989b9024b031daa955575a271560fa1c0ba4af2965a081a1c08
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [13]:
from apyori import apriori

In [17]:
list(apriori(transactions,min_support = 0.5))

[]

In [18]:
list(apriori(transactions,min_support = 0.1))

[]

In [19]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [24]:
frequentItemsets = list(apriori(transactions,min_support = 0.03))
frequentItemsets

[RelationRecord(items=frozenset({'20712'}), support=0.033667953667953665, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20712'}), confidence=0.033667953667953665, lift=1.0)]),
 RelationRecord(items=frozenset({'20719'}), support=0.03254826254826255, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20719'}), confidence=0.03254826254826255, lift=1.0)]),
 RelationRecord(items=frozenset({'20724'}), support=0.04054054054054054, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20724'}), confidence=0.04054054054054054, lift=1.0)]),
 RelationRecord(items=frozenset({'20725'}), support=0.062084942084942085, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.062084942084942085, lift=1.0)]),
 RelationRecord(items=frozenset({'20726'}), support=0.04003861003861004, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozense

In [25]:
len(frequentItemsets)

80

In [26]:
frequentItemsets[-1]

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [27]:
frequentItemsets[-1][0]

frozenset({'22697', '22699'})

In [36]:
len(frequentItemsets[-1][0])

2

In [37]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if len(i[0])>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216,
ordered_statistics=
[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}),items_add=frozenset({'85099B'}),
confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}),
confidence=0.39016393442622943, lift=8.20897311262335)])


In [39]:
data[data['StockCode']=='85099B']

InvoiceNo StockCode              Description  Quantity  \
177       536386    85099B  JUMBO BAG RED RETROSPOT       100   
234       536390    85099B  JUMBO BAG RED RETROSPOT       100   
533       536409    85099B  JUMBO BAG RED RETROSPOT         2   
790       536464    85099B  JUMBO BAG RED RETROSPOT         1   
1069      536522    85099B  JUMBO BAG RED RETROSPOT         1   
...          ...       ...                      ...       ...   
538603    581412    85099B  JUMBO BAG RED RETROSPOT         3   
538913    581421    85099B  JUMBO BAG RED RETROSPOT         9   
539338    581439    85099B  JUMBO BAG RED RETROSPOT         1   
540434    581486    85099B  JUMBO BAG RED RETROSPOT        10   
541536    581498    85099B  JUMBO BAG RED RETROSPOT         5   

             InvoiceDate  UnitPrice  CustomerID         Country  
177     12/01/2010 09:57       1.65     16029.0  United Kingdom  
234     12/01/2010 10:19       1.65     17511.0  United Kingdom  
533     12/01/2010 11:45       1.95     17908.0  United Kingdom  
790     12/01/2010 12:23       1.95     17968.0  United Kingdom  
1069    12/01/2010 12:49       1.95     15012.0  United Kingdom  
...                  ...        ...         ...             ...  
538603  12/08/2011 14:38       2.08     14415.0  United Kingdom  
538913  12/08/2011 15:23       2.08     15898.0  United Kingdom  
539338  12/08/2011 16:30       4.13         NaN  United Kingdom  
540434  12/09/2011 09:38       2.08     17001.0  United Kingdom  
541536  12/09/2011 10:26       4.13         NaN  United Kingdom  

[2159 rows x 8 columns]

In [38]:
data[data['StockCode']=='22386']

InvoiceNo StockCode              Description  Quantity  \
86        536378     22386  JUMBO BAG PINK POLKADOT        10   
789       536464     22386  JUMBO BAG PINK POLKADOT         1   
1428      536542     22386  JUMBO BAG PINK POLKADOT        30   
2427      536591     22386  JUMBO BAG PINK POLKADOT         1   
2712      536592     22386  JUMBO BAG PINK POLKADOT         3   
...          ...       ...                      ...       ...   
534742    581167     22386  JUMBO BAG PINK POLKADOT        20   
537138    581238     22386  JUMBO BAG PINK POLKADOT         5   
539585    581439     22386  JUMBO BAG PINK POLKADOT         2   
541421    581498     22386  JUMBO BAG PINK POLKADOT         1   
541806    581579     22386  JUMBO BAG PINK POLKADOT        10   

             InvoiceDate  UnitPrice  CustomerID         Country  
86      12/01/2010 09:37       1.95     14688.0  United Kingdom  
789     12/01/2010 12:23       1.95     17968.0  United Kingdom  
1428    12/01/2010 14:11       1.95     16456.0  United Kingdom  
2427    12/01/2010 16:57       1.95     14606.0  United Kingdom  
2712    12/01/2010 17:06       4.21         NaN  United Kingdom  
...                  ...        ...         ...             ...  
534742  12/07/2011 14:52       2.08     13534.0  United Kingdom  
537138  12/08/2011 10:53       4.13         NaN  United Kingdom  
539585  12/08/2011 16:30       4.13         NaN  United Kingdom  
541421  12/09/2011 10:26       4.13         NaN  United Kingdom  
541806  12/09/2011 12:19       1.79     17581.0  United Kingdom  

[1251 rows x 8 columns]

# HW 2 หา pattern ที่น่าสนใจใน Dataset หนังอินเดีย (india imdb)